isvcs/container.go

```diff
@@ -459,17 +469,61 @@ func (svc *IService) run() {
 					req.response <- err
 					continue
 				}
+
+				if !collecting {
+					go svc.stats(haltStats)
+					go svc.doHealthChecks(haltHealthChecks)
+					collecting = true
+				}
 				req.response <- nil
 			}
 		case rc := <-svc.exited:
-			glog.Errorf("isvc %s unexpectedly received exit code %d", svc.Name, rc)
-			if svc.exited, err = svc.start(); err != nil {
-				glog.Errorf("Error restarting isvc %s: %s", svc.Name, err)
+			svc.setStoppedHealthStatus(ExitError(rc))
+			glog.Errorf("isvc %s exited unexpectedly; rc=%d", svc.Name, rc)
+
+			stopService := svc.isFlapping()
+			if !stopService {
+				glog.Infof("Restarting isvc %s ", svc.Name)
+				if svc.exited, err = svc.start(); err != nil {
+					glog.Errorf("Error restarting isvc %s: %s", svc.Name, err)
+					stopService = true
+				}
+			}
+
+			if stopService {
+				if collecting {
+					haltStats <- struct{}{}
+					if len(svc.HealthChecks) > 0 {
+						haltHealthChecks <- struct{}{}
+					}
+					collecting = false
+				}
+				glog.Errorf("isvc %s not restarted; failed too many times", svc.Name)
+				svc.stop()
+				svc.exited = nil
 			}
 		}
 	}
 }
 
+// Return true if the service is flapping (starting/stopping repeatedly).
```

Comment:
perhaps "isFlapping returns true if the service is flapping (starting/stopping repeatedly)."